Primarily customisability enhancements

sergv opened the pull request:
I would like propose a bunch of improvements, which can be roughly summarised as follows:
- Performance enhancements - emacs would do less work and would (hopefully) work faster. The only danger here is that when I changed `string-match` to `string-match-p` and `looking-at` to `looking-at-p` I may have missed a place where match data was subsequently accessed. But hopefully that didn't happen...
- Customisability enhancements - move more stuff (i.e. keymap for `intero-multiswitch`) to variables or add new variables (i.e. `intero-extra-ghc-options` and `indero-extra-ghci-options`) so that a user may influence intero's behaviour without having to change it's source code.
- New feature - completion in repl for `:set` command. Could probably be improved to complete other commands as well, but I'd rather add this small bit now and come up with a followup.
- Windows bug fix - `intero-paths-for-same-file` does not do the right thing on Windows platform whet fed arguments like `C:/foo/bar/Baz.hs` and `C:\foo\bar\Baz.hs`.

acowley commented on elisp/intero.el:
```diff
@@ -920,6 +939,23 @@ considered."
           (not (looking-at-p (rx (| space line-end)))))))))
 
+(defun intero-completions-grab-ghci-command ()
+  "Grapb prefix for a ghci command like :set.
```
s/Grapb/Grab/

acowley commented:
Does the [switch back to using the GHCi repl](https://github.com/commercialhaskell/intero/commit/f80d503fd682defc750406ef3021091b732edffe) impact the approach to expanding autocompletion of repl commands?

acowley commented:
The `-p` changes look good to me, but I haven't tried it out yet.

sergv commented:
Regarding 'switch back to using the GHCi repl'' it seems that at some point there might be a difference between the `intero` and the `ghci` executables. However, I expect user-visible commands like `:set` to be largely compatible.

purcell commented on elisp/intero.el:
```diff
@@ -518,7 +534,7 @@ If the problem persists, please report this as a bug!")))
                           (looking-back "$(" 1))
                         (+ 2 (current-column))
-                      (if (looking-at "$(")
+                      (if (looking-at-p "$(")
```
This is certainly a reasonable change semantically, but I'd be surprised if it affects performance. After all, `looking-at-p` is defined as
```
(let ((inhibit-changing-match-data t))
    (looking-at regexp))
```

sergv replied:
Well, the C implementation of `looking-at` does pay attention to this variable and does more work if it's `nil`. For instance, please check https://github.com/emacs-mirror/emacs/blob/master/src/search.c#L334 (and possibly implementation of `re_match_2` as well).

purcell replied:
Thanks, great to know.

purcell commented on elisp/intero.el:
```diff
@@ -796,19 +812,20 @@ CHECKER and BUFFER are added to each item parsed from STRING."
+                                         (string-to-number (match-string 3 string)))
+                                :col2 (when (match-string 5 string)
+                                        (string-to-number (match-string 5 string)))))))
+      span)))
```
+1 for this change. (The whole thing could likely be written more efficiently, by breaking the loop when a regexp matches, rather than continuing to (potentially) match other regexps.)

sergv replied:
I think there's even more performant implementation: just test all three regexes at once by single `match-string`. Regexes modify the same groups for file name, line number and column - single match should populate them properly.

purcell commented on elisp/intero.el:
```diff
@@ -893,7 +910,7 @@ Other arguments are IGNORED."
                          (cl-remove-if-not
                           (lambda (candidate)
-                            (string-match (concat "^" prefix) candidate))
+                            (string-match-p (concat "^" prefix) candidate))
```
If `prefix` isn't a regexp, which it shouldn't be, this could/should simply be `string-prefix-p`.

purcell commented on elisp/intero.el:
```diff
@@ -904,6 +921,8 @@ If specified, MINLEN is the shortest completion which will be
 considered."
   (when (intero-completions-can-grab-prefix)
     (let ((prefix (cond
+                   ((and (eq 'intero-repl-mode major-mode)
```
Prefer `(derived-mode-p 'intero-repl-mode)` to testing for exact equality of `major-mode`.

purcell replied:
Generally you'd put the test for applicability (i.e. the major mode test here) in the function which grabs the prefix: it would return nil if inapplicable.

purcell commented on elisp/intero.el:
```diff
@@ -2339,6 +2377,13 @@ a list is returned instead of failing with a nil result."
+    (define-key map (kbd "C-c C-k") 'abort-recursive-edit)
+    (define-key map (kbd "C-g")     'abort-recursive-edit)
+    map))
+
```
Nice.

sergv commented:
I have fixed test for `intero-repl-mode` and `string-prefix-p` over `string-match-p` for pragmas. I have also added another optimization where multiple regexps are now tested in one go through single call to `string-match`.

purcell commented on elisp/intero.el:
```diff
@@ -747,14 +763,19 @@ CHECKER and BUFFER are added to each item parsed from STRING."
+                                     "["
+                                     (or "-Wdeferred-type-errors"
+                                         "-Wdeferred-out-of-scope-variables"
+                                         "-Wtyped-holes")
```
Really nice.
